### ISSUE TITLE
Elasticsearch: query_string query prioritizes synonyms

### ISSUE THREAD

**Issue by ERosendo** [freelawproject/courtlistener#4816]
While working on PR #4814 , I tried searching for cases with cause 31:3730 Qui Tam False Claims Act using both the main search bar and the 'Refine Your Query' cause filter. The main search bar results included the following case:

![Image](https://github.com/user-attachments/assets/ff957b04-fc2b-4806-abcc-d6b7ef7b3af2)

However, using the specific cause filter yielded no results. The screenshot reveals that Elasticsearch matches 'FCA'(a synonym) but not 'False Claims Act'. This explains the lack of results when using the cause filter, as 'FCA' isn't indexed in the 'cause' field

**Comment by albertisfu** [comment]
Well actually the problem primary lies in the `search_analyzer` because we decided to not add synonyms at indexing time so synonyms are only applied during search.

The problem is related to how synonyms in the query are expanded:

`31:3730 Qui Tam False Claims Act`

The terms remain in the same positions up to "False." However, as shown in the image below, the position of "claim" differs: at indexing time, its position is 5, while at search time, it is 7. This discrepancy causes the query to mismatch, preventing the document from being returned as a result.

The synonyms for these terms are defined as follows:
`false claims act, federal communications act, fca`

The issue seems to stem from how synonym term expansions are handled specially on synonyms with multiple definitions, resulting in the following expansions:

- `false claims act`
- `federal communications act`
- `fca`
- `false claims act -> federal communications act`
- `false claims act -> fca`
- `federal communications act -> false claims act`
- `federal communications act -> fca`
- `fca -> false claims act`
- `fca -> federal communications act`

In practice, this adds new terms in different positions.

The first term of each synonym retains its original position:

- `false, feder, fals: 4`

However, additional terms are inserted at different positions:

- `commun: 5`
- `act: 6`
- `claim: 7`

This positional shift of "claim" creates a problem.

Disabling synonyms can resolve the issue but it'll cause the query to only match the synonyms and ignore the original term, which is not expected in most cases.

It may be worth it to investigate if adjusting the position of the synonym filter in the search analyzer could resolve this issue. 

Another option might be to disable synonyms  in filters.


![Image](https://github.com/user-attachments/assets/2e88df97-e55d-4cdc-9ea4-639b0b96395d)

**Comment by albertisfu** [comment]
> If we enable synonyms for the description field and re-index, does that fix this?

Yes, I tested it, and it does solve the problem.

However, I recall two main reasons why we chose not to index synonyms initially:
- It could significantly increase the index size.
- Every time we need to introduce new synonyms, a complete re-index is required.

During my test, I discovered an additional issue with the following query:

`31:3730 Qui Tam False Claims Act`

After indexing synonyms, this query still did not match the desired result. However, it worked after removing `31:`.

The issue is that `31:3730` is being interpreted as a fielded search, with `31` being treated as a field. Since `31` is not recognized as a field in the index mapping, the query fails.

This problem does not occur in the search box because `31:3730` is detected as a docket number. It is enclosed in quotes, becoming `"31:3730"`, which ensures the search is treated as exact and not as a fielded search. To address this, we can apply for filters the same preprocessing of terms performed by the `cleanup_main_query` function.


Back to the Synonyms Issue I tried a couple of approaches:

1. **Removing Stemming from the Analyzers:**
    The problem persisted, indicating that the combination of synonyms and stemming is not the issue so changing the order of these filters won't help.
    
2. **Testing with Synonym Variations:**
    - Indexed synonyms as:
        - `false claims act, fca` This worked, the document was matched.
        - `false claims act, federal communications lorem, fca`This also worked. Changing `act` from `federal communications act` to a different term fixed the issue.

From this, this seems like a be a bug in the synonym expansion process to me. It appears that there is a recursive replacement of terms during expansion that may cause the positions of certain terms (like claim in our example query) to switch and cause the query to not match the document.

In brief this synonym expansion issue affects cases where:
There are more than two definitions for a synonym and some terms are repeated across multiple definitions, as in this example:

`false claims act, federal communications act, fca`

**Comment by mlissner** [comment]
Got it, so it feels like disabling synonyms is the way to go?

**Comment by albertisfu** [comment]
Yes, disable synonyms in text filters and also include `cleanup_main_query` as a preprocessing step in the filters to solve the other issue described.

However, this issue is also present in search box queries. Its impact is less noticeable, though, since the query string in the search box matches multiple fields simultaneously, as shown in Eduardo's screenshot in the issue description

**Comment by mlissner** [comment]
I was afraid you'd say that. And there's really nothing we can do there, right? Solr had similar bugs with synonyms....

**Comment by albertisfu** [comment]
> And there's really nothing we can do there, right?

I tested a few possible alternatives here:

- Using the [multiplexer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-multiplexer-tokenfilter.html) filter looked promising since we could use it to place the synonym terms and the original terms in the same position. Unfortunately, it doesn't work well with multi-word synonyms.
- Changing from the Synonym Graph filter to the simple Synonym filter solved the issue. However, using the simple Synonym filter caused other tests related to multi-word synonyms to fail because the simple Synonym filter performs poorly with multi-word synonyms.
- Since the issue is directly related to the Synonym Graph filter and its [limitations](https://blog.mikemccandless.com/2012/04/lucenes-tokenstreams-are-actually.html) due to its graph representation, I tried using the [Flatten Graph](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-flatten-graph-tokenfilter.html) filter after applying the Synonym Graph filter. Using this filter resolved the position problem in the graph representation and matched the documents correctly. Additionally, all other tests related to synonyms passed successfully.

A more carefully revision to `flatten_graph` filter might be required to ensure it doesn't disrupt other types of synonym matching, but so far, it appears to be the solution. To implement this, we need to add `flatten_graph` after `synonym_filter` in the `search_analyzer`:

```
     "search_analyzer": {
                "type": "custom",
                "tokenizer": "whitespace",
                "filter": [
                    "lowercase",
                    "synonym_filter",
                    "flatten_graph",
                    "custom_word_delimiter_filter",
                    "remove_leading_zeros",
                    "english_stop",
                    "english_stemmer",
                    "remove_duplicates",
                ],
            }
```

It'll be necessary to close the indices, update the analyzers for each index, and then reopen them so that this change takes effect.

**Comment by mlissner** [comment]
I'm suspicious that we will find more tests and failing queries soon, but this looks like a great first step! Let's give it a try.

**Comment by mlissner** [comment]
Y'all seem to be working on this together. Who should we assign it to? We've got this, a [PR by Eduardo](https://github.com/freelawproject/courtlistener/pull/4814) and an [issue filed by Alberto but assigned to Eduardo](https://github.com/freelawproject/courtlistener/issues/3993).

Thoughts? Do we put it in this sprint, or no?

**Comment by ERosendo** [comment]
Hey @mlissner I'm happy to keep working on this issue and I think I can fit it into this sprint.

**Comment by ERosendo** [comment]
Hey @albertisfu, I've encountered discrepancies between the documented behavior and actual performance of the [Flatten Graph](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-flatten-graph-tokenfilter.html) filter. I've filed an [issue](https://github.com/elastic/elasticsearch/issues/119175) on the ES repository to highlight this problem. Due to these inconsistencies, I think it's best to hold off on using it for now. We should wait for the issue to be reviewed to avoid potential unintended consequences.

**Comment by albertisfu** [comment]
Thanks @ERosendo I agree that it's better to clarify the expected behavior of the Flatten filter before adding it to our search analyzer.

**Comment by mlissner** [comment]
@ERosendo, does the bug you found create a bug in our system?

**Comment by ERosendo** [comment]
>  does the bug you found create a bug in our system?

Apologies, I realize I already addressed this question in a related issue. I'm not sure how my comment ended up there – perhaps I clicked on the wrong card while navigating the sprint board. You can find my comment here:

https://github.com/freelawproject/courtlistener/issues/3993#issuecomment-2557588195

**Comment by mlissner** [comment]
Is there any hope of doing something here or did we conclude it's a lost cause? If it's a lost cause, should we move it to done and just keep the issue around for future reference?

**Comment by ERosendo** [comment]
I just checked the status of the [issue](https://github.com/elastic/elasticsearch/issues/119175) we submitted in the Elasticsearch repository. I noticed they've assigned a developer to work on it and set a priority. I think we can keep this issue around and keep an eye on it for updates.

**Comment by mlissner** [comment]
Great. I'll wait for word from you that that issue is fixed and that we should upgrade Elastic (yay!), and then fix this too. Thanks!
